### PR TITLE
varianter: Make it properly pluggable [v4]

### PR DESCRIPTION
**Pull request by ldoktor**
This PR is inspired by https://github.com/avocado-framework/avocado/pull/1591 and is the next stage of https://github.com/avocado-framework/avocado/pull/1685 . First it cleans the current usage of varianter and then it turns it into a proper plugin defined in `avocado.core.plugin_interfaces`. Along the way there are several fixes and enhancements and at the end there is the promised rewrite of the multiplexer documentation, which took me several days to put together and I hope it's understandable and clear enough.

v1: https://github.com/avocado-framework/avocado/pull/1716
v2: https://github.com/avocado-framework/avocado/pull/1763
v3: https://github.com/avocado-framework/avocado/pull/1812

Changes:

```yaml
v2: Just rebased to see whether the travis failures were false-negatives
v3: Rebase to latest master
v3: Using `to_str` every since the beginning
v3: Use the `args.contents` to respect verbosity in `avocado.multiplex`
v3: Improved commit message of the "Make it properly pluggable" commit
v3: Squashed all the proper-pluggin-related commits including stevedore code
v3: Use human friendly "enum" to define multiplex verbosity
v3: Move the docstring fixes to where they were introduced (if introduced)
v3: Added unittest for the fixed skip_new_data_check
v3: Use **kwargs rather than extra to pass optional arguments to to_str
v3: Some minor refinements
v4: Moved several changes earlier to avoid double-changes
v4: Use :meth:, :class:, :mod: properly in docstrings
v4: Several language improvements (English)
v4: New commit to avoid the E1101 pylint warning
v4: Reworked the --summary and --variants defaults handling
v4: Replaced the "Fix the return value in skip_new_data_check" with commit which removes
    it as it became unnecessary. The unittest is still valid as it crashes on older versions.
v4: Rebased
```

Not changed:
```yaml
v4: Changes proposed elsewhere than inside the PR (it's really hard to trace all the changes)
v4: The invoke_kwds (not sure how, args are not available on __init__ time)
v4: Varianter.__len__ - it depends on the test length, Varianter length is not really needed
v4: Few other things which were not clarified (answered) before creating v4
```

### CODE REVIEW

File: avocado/core/varianter.py [hunk 1]
```diff
@@ -519,11 +525,10 @@ def get_number_of_tests(self, test_suite):
...
         # Currently number of tests is symmetrical
-        if self.variants:
-            no_variants = sum(1 for _ in self.variants)
-            if no_variants > 1:
-                self._has_multiple_variants = True
-            return (len(test_suite) * no_variants)
+        no_variants = len(self.variants)
+        if no_variants > 1:
+            self._has_multiple_variants = True
+            return len(test_suite) * no_variants
```
Comment (clebergnu) [inline avocado/core/varianter.py]: OK.

File: avocado/core/varianter.py [hunk 2]
```diff
@@ -432,6 +434,6 @@ def itertests(self):
             for variant in iter(iter_variants):
                 yield variant
         else:   # No variants, use template
-            yield {"variant": self.default_params.get_leaves(),  # self.default_params is TreeNode pylint: disable=E1101
+            yield {"variant": self._default_params.get_leaves(),
```
Comment (clebergnu) [inline avocado/core/varianter.py]: You asked on the previous version, so here is the answer again: please include (and squash) this one.

File: avocado/core/varianter.py [hunk 3]
```diff
@@ -525,24 +534,31 @@ def get_number_of_tests(self, test_suite):
...
-            if self._has_multiple_variants:
-                for i, variant in enumerate(self.variants, 1):
-                    yield i, (variant, self._mux_path)
-            else:
-                yield None, (iter(self.variants).next(), self._mux_path)
+            for i, variant in enumerate(self.variants, 1):
+                yield i, (variant, self._mux_path)
         else:   # No variants, use template
-            yield None, None
+            yield None, (self.default_params.get_leaves(), self._mux_path)
```
Comment (clebergnu) [inline avocado/core/varianter.py]: OK (don't forget about 2a7279ffc3eb76b06573de8a373be927f8c963f7 though)

File: avocado/plugins/multiplex.py [hunk 4]
```diff
@@ -63,34 +86,34 @@ def run(self, args):
...
-        for line in variants.splitlines():
+                summary += 2
+        else:
+            if args.contents:
+                variants += 2
+
+        # Produce the output
+        lines = args.avocado_variants.to_str(summary=summary,
+                                             variants=variants,
+                                             use_utf8=use_utf8)
```
Comment (clebergnu) [inline avocado/plugins/multiplex.py]: This is a fix that was spotted as needed on b54181a17cc2b21e471cde812434d766886d4f11.

File: avocado/plugins/multiplex.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def configure(self, parser):
                             help="Verbosity of the list of variants. " +
                             verbosity_levels, default=1)
         parser.add_argument('--system-wide', action='store_false',
-                            default=True, dest="mux-skip-defaults",
+                            default=True, dest="variants-skip-defaults",
```
Comment (clebergnu) [inline avocado/plugins/multiplex.py]: OK.

File: selftests/functional/test_replay_basic.py [hunk 6]
```diff
@@ -197,6 +197,16 @@ def test_run_replay_fallbackdir(self):
         expected_rc = exit_codes.AVOCADO_ALL_OK
         self.run_and_check(cmd_line, expected_rc)
 
+    def test_run_replay_and_mux(self):
```
Comment (clebergnu) [inline selftests/functional/test_replay_basic.py]: OK.

File: avocado/core/mux.py [hunk 7]
```diff
@@ -116,7 +116,9 @@ def __iter__(self):
         if self.root is None:
             return
         for i, variant in enumerate(self.variants, 1):
-            yield i, (variant, self.mux_path)
+            yield {"variant_id": i,
```
Comment (clebergnu) [inline avocado/core/mux.py]: OK.

File: docs/source/index.rst [hunk 8]
```diff
@@ -15,7 +15,7 @@ Contents:
    Loaders
    LoggingSystem
    Sysinfo
-   Mux
+   TestParameters
```
Comment (clebergnu) [inline docs/source/index.rst]: Yes, these changes look good.

File: avocado/core/varianter.py [hunk 9]
```diff
@@ -361,7 +361,7 @@ def _process_default_params(self, args):
 
     def is_parsed(self):
         """
-        Reports whether the tree was already multiplexed
+        Reports whether the varianter was already parsed
```
Comment (clebergnu) [inline avocado/core/varianter.py]: OK.

File: avocado/core/varianter.py [hunk 10]
```diff
@@ -333,7 +333,7 @@ def __init__(self, debug=False):
 
     def parse(self, args):
         """
-        Apply options defined on the cmdline
+        Apply options defined on the cmdline and initialize the plugins.
```
Comment (clebergnu) [inline avocado/core/varianter.py]: OK.

File: avocado/plugins/multiplex.py [hunk 11]
```diff
@@ -82,29 +82,15 @@ def run(self, args):
...
-                    continue
-                fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
-                for record in sorted(env):
-                    log.debug(fmt, *record)
+        if args.mux_debug:
+            # In this version `avocado_variants.debug` is not set properly,
+            # let's force-enable it before calling to_str to
+            # get the expected results.
+            args.avocado_variants.debug = True
+        use_utf8 = settings.get_value("runner.output", "utf8",
```
Comment (ldoktor) [inline avocado/plugins/multiplex.py]: True for line 90, it is not needed as `variants` does not support utf8. On line 79 it's used so I'm keeping it.

**Pull request closed without merging**